### ISSUE TITLE
Errors 'Failed to parse available descriptor chain: DescriptorChainTooShort' when executing OSv ZFS image

### ISSUE THREAD

**Issue by wkozaczuk** [firecracker-microvm/firecracker#857]
First of I wanted to congratulate on this great new VMM project. I have been very impressed by super fast startup (under 5ms), REST API, excellent diagnostics and small code base.

In last 2-3 week I have been able to enhance OSv to run on firecracker. [OSv](https://github.com/cloudius-systems/osv) is a unikernel that implements minimal subset of POSIX ABI (no fork, etc) in order to run unmodified Linux applications (please read more [here](https://www.mikelangelo-project.eu/technology/universal-unikernel-osv/)). 

I am pleased to inform that based on some preliminary tests OSv is able to boot in 7-10 ms on firecracker and execute simple hello world C app in 10-15 ms on average (time elapsed from the moment reported when receiving "InstanceStart" API request to 'VMM stopping"). Please see log excerpts below.

```
OSv v0.52.0-25-gc0d4f463
Cmdline: --bootchart /hello virtio_mmio.device=4K@0xd0000000:5
	disk read (real mode): 0.00ms, (+0.00ms)
	uncompress lzloader.elf: 0.00ms, (+0.00ms)
	TLS initialization: 1.49ms, (+1.49ms)
	.init functions: 2.68ms, (+1.19ms)
	SMP launched: 4.18ms, (+1.50ms)
	VFS initialized: 4.91ms, (+0.73ms)
	Network initialized: 5.30ms, (+0.39ms)
	pvpanic done: 6.30ms, (+1.00ms)
	pci enumerated: 6.30ms, (+0.00ms)
	drivers probe: 6.34ms, (+0.04ms)
	drivers loaded: 6.67ms, (+0.33ms)
	ROFS mounted: 6.83ms, (+0.16ms)
	Total time: 6.83ms, (+0.00ms)
Hello from C code
```

```
2019-01-16T22:05:19.868682846 [anonymous-instance:INFO:api_server/src/http_service.rs:615] The API server received a synchronous Put request on "/actions" with body "{n "action_type": "InstanceStart"n }".
2019-01-16T22:05:19.868697369 [anonymous-instance:INFO:vmm/src/lib.rs:1145] VMM received instance start command
2019-01-16T22:05:19.868751950 [anonymous-instance:INFO:vmm/src/vstate.rs:97] Guest memory starts at 7fc485800000
2019-01-16T22:05:19.872845198 [anonymous-instance:INFO:api_server/src/http_service.rs:581] The synchronous Put request on "/actions" with body "{n "action_type": "InstanceStart"n }" was executed successfully. Status code: 204 No Content.
2019-01-16T22:05:19.880848783 [anonymous-instance:INFO:vmm/src/lib.rs:1204] Vmm is stopping.
```

Even though I have been able to verify that OSv works with both block and networking devices in all experiments I have conducted, there are specific cases when I see firecracker report ** Failed to parse available descriptor chain: DescriptorChainTooShort** errors in its log. I have narrowed it down to the scenarios when OSv mounts ZFS filesystem as opposed to simple ROFS (read-only) filesystem to load an application from to execute. Please see the log snippets below.

```
2019-01-16T22:19:39.836369636 [anonymous-instance:INFO:api_server/src/http_service.rs:615] The API server received a synchronous Put request on "/actions" with body "{n "action_type": "InstanceStart"n }".
2019-01-16T22:19:39.836400348 [anonymous-instance:INFO:vmm/src/lib.rs:1145] VMM received instance start command
2019-01-16T22:19:39.836463219 [anonymous-instance:INFO:vmm/src/vstate.rs:97] Guest memory starts at 7fc24a800000
2019-01-16T22:19:39.840775858 [anonymous-instance:INFO:api_server/src/http_service.rs:581] The synchronous Put request on "/actions" with body "{n "action_type": "InstanceStart"n }" was executed successfully. Status code: 204 No Content.
2019-01-16T22:19:39.887184611 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.887391996 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.902476667 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.902631780 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.904698273 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.904852813 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.906634448 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.906774302 [anonymous-instance:ERROR:devices/src/virtio/block.rs:309] Failed to parse available descriptor chain: DescriptorChainTooShort
2019-01-16T22:19:39.906855031 [anonymous-instance:INFO:vmm/src/lib.rs:1204] Vmm is stopping.
```

What is interesting I could not see any bad side effects on OSv side but I would like to understand what is the nature of these errors. I am suspecting it most likely caused by some bugs in OSv virtio-mmio implementation which I had to code to make OSv to boot on firecracker. 

Would you please explain what exactly 'Failed to parse available descriptor chain: DescriptorChainTooShort' means and ideally what possibly I am doing wrong on guest side?

One clue might be that occasionally I see firecracker return non-VIRTIO_MMIO_INT_VRING value when ack-ing an interrupt on OSv side (please see this interrupt ack-ing code in OSv - https://github.com/wkozaczuk/osv/blob/fire_cracker_rfc/drivers/virtio-mmio.cc#L84-L94). I wonder if that is related.

**Comment by andreeaflorescu** [comment]
Hey, nice work on OSv! We will take a look and let you know about our findings. It would help if you could provide the filesystem and the boot source along with the API calls you are using to setup Firecracker so we can reproduce this locally.

**Comment by dhrgit** [comment]
@wkozaczuk Just an idea:
Firecracker issues `DescriptorChainTooShort` when a blk request fails to provide a status descriptor (immediately following the data descriptor). If your code only gets this error when using writable file systems, perhaps there's a problem causing (some) VIRTIO_BLK_T_OUT requests to be sent without a status descriptor?

**Comment by andreeaflorescu** [comment]
I am not very sure that it comes from the status descriptor. It can also be an error on the data descriptor. In Firecracker we throw the same error for both status & data descriptor errors. We should probably differentiate somehow between the two. We should also probably add some debug prints that can tell us more about the root cause of the problem.

**Comment by wkozaczuk** [comment]
These are the artifacts you will need to reproduce the issue (I pushed them to my S3 bucket which hopefully you will be able to wget from):
- [OSv kernel](https://wkozaczuk-osv-images.s3.amazonaws.com/firecracker/loader.elf)
- [ROFS filesystem image with hello world app - no errors](https://wkozaczuk-osv-images.s3.amazonaws.com/firecracker/hello.rofs)
- [ZFS filesystem image with hello world app - WITH errors](https://wkozaczuk-osv-images.s3.amazonaws.com/firecracker/hello.zfs)

Here are the scripts you will need to run OSv:
```
curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT 'http://localhost/drives/rootfs' \
    -H 'Accept: application/json'           \
    -H 'Content-Type: application/json'     \
    -d '{
        "drive_id": "rootfs",
        "path_on_host": "<folder_to_where_to_downloaded_from>/hello.rofs", //Or hello.zfs
        "is_root_device": false,
        "is_read_only": false
    }'

curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT 'http://localhost/boot-source'   \
    -H 'Accept: application/json'           \
    -H 'Content-Type: application/json'     \
    -d '{
        "kernel_image_path": "<folder_to_where_to_downloaded_from>/loader.elf",
        "boot_args": "/hello"
    }'

curl --unix-socket /tmp/firecracker.socket -i \
    -X PUT 'http://localhost/actions'       \
    -H  'Accept: application/json'          \
    -H  'Content-Type: application/json'    \
    -d '{
        "action_type": "InstanceStart"
     }'
```

You should NOT see any errors with hello.rofs and see aforementioned errors in firecracker logs when running hello.zfs. It was working as of this commit - https://github.com/firecracker-microvm/firecracker/commit/0a363fc3af98830a64f7a8b5b331db6ecf772fc9.

**Comment by dhrgit** [comment]
@wkozaczuk It looks like the errors are caused by unexpected flush requests.

The Firecracker virtio-blk device doesn't advertise (nor support) VIRTIO_BLK_F_FLUSH. However, it appears the guest-side driver ignores this, and still sends flush requests.

On our side, we should:
- report these errors more accurately (e.g. unsupported requests)
- add VIRTIO_BLK_F_FLUSH support

We'll be opening new issues to track that.

**Comment by wkozaczuk** [comment]
@dhrgit Thanks for this find. I will make corrections on OSv side to detect if VIRTIO_BLK_F_FLUSH is supported and only then send flush requests.

**Comment by wkozaczuk** [comment]
I think there may be a problem in handling of VIRTIO_BLK_T_FLUSH in this piece of code - https://github.com/firecracker-microvm/firecracker/blob/c56cae3a310d909ab625526e182996f6d84f441d/devices/src/virtio/block.rs#L171-L173

Per this from virtio spec:
> 5.2.6.1 Driver Requirements: Device Operation
A driver MUST NOT submit a request which would cause a read or write beyond capacity.
A driver SHOULD accept the VIRTIO_BLK_F_RO feature if offered.
A driver MUST set sector to 0 for a VIRTIO_BLK_T_FLUSH request. A driver SHOULD NOT include any data in a VIRTIO_BLK_T_FLUSH request.

driver should NOT include any data and Request::parse() expect data even for VIRTIO_BLK_T_FLUSH. Is this correct?

**Comment by dhrgit** [comment]
Yes, there are problems in the virtio-blk code. E.g. currently, Firecracker doesn't handle VIRTIO_BLK_T_FLUSH requests properly. Since it isn't advertising flush support, it should respond by setting the status to VIRTIO_BLK_S_UNSUPP.

**Comment by andreeaflorescu** [comment]
Should this issue be fixed by #893?

**Comment by dhrgit** [comment]
> Should this issue be fixed by #893?

Yes. I've also confirmed the fix by testing with the fs and OSv kernel provided by @wkozaczuk . Closing this.

**Comment by wkozaczuk** [comment]
I have also run tests on my side independently and I am pleased to report it is fixed. Thanks for fixing it!
